### PR TITLE
check if AbstractMatrix supports Tables.istable in a constructor

### LINKED ISSUE #2433: Deprecate `DataFrame(::AbstractMatrix)`
I feel like it's rare that folks actually want to call this constructor without specifying column names. Also, it isn't that hard to do this with the tables interface via `DataFrame(Tables.table(X))`. 

```
julia> DataFrame(Tables.table(x))
10×5 DataFrame
│ Row │ Column1  │ Column2  │ Column3   │ Column4  │ Column5   │
│     │ Float64  │ Float64  │ Float64   │ Float64  │ Float64   │
├─────┼──────────┼──────────┼───────────┼──────────┼───────────┤
│ 1   │ 0.829388 │ 0.100101 │ 0.733667  │ 0.591796 │ 0.538582  │
│ 2   │ 0.573906 │ 0.402706 │ 0.41509   │ 0.899583 │ 0.482383  │
│ 3   │ 0.703773 │ 0.729495 │ 0.375752  │ 0.782216 │ 0.425253  │
│ 4   │ 0.5717   │ 0.93305  │ 0.0922584 │ 0.211547 │ 0.123723  │
│ 5   │ 0.927349 │ 0.654163 │ 0.11306   │ 0.452749 │ 0.921814  │
│ 6   │ 0.514157 │ 0.891617 │ 0.481489  │ 0.368915 │ 0.718185  │
│ 7   │ 0.717906 │ 0.627349 │ 0.371058  │ 0.958308 │ 0.925878  │
│ 8   │ 0.104714 │ 0.995778 │ 0.671143  │ 0.410917 │ 0.547495  │
│ 9   │ 0.373802 │ 0.147139 │ 0.425674  │ 0.893018 │ 0.937846  │
│ 10  │ 0.26242  │ 0.30546  │ 0.750857  │ 0.196515 │ 0.0470533 │

julia> DataFrame(x)
10×5 DataFrame
│ Row │ x1       │ x2       │ x3        │ x4       │ x5        │
│     │ Float64  │ Float64  │ Float64   │ Float64  │ Float64   │
├─────┼──────────┼──────────┼───────────┼──────────┼───────────┤
│ 1   │ 0.829388 │ 0.100101 │ 0.733667  │ 0.591796 │ 0.538582  │
│ 2   │ 0.573906 │ 0.402706 │ 0.41509   │ 0.899583 │ 0.482383  │
│ 3   │ 0.703773 │ 0.729495 │ 0.375752  │ 0.782216 │ 0.425253  │
│ 4   │ 0.5717   │ 0.93305  │ 0.0922584 │ 0.211547 │ 0.123723  │
│ 5   │ 0.927349 │ 0.654163 │ 0.11306   │ 0.452749 │ 0.921814  │
│ 6   │ 0.514157 │ 0.891617 │ 0.481489  │ 0.368915 │ 0.718185  │
│ 7   │ 0.717906 │ 0.627349 │ 0.371058  │ 0.958308 │ 0.925878  │
│ 8   │ 0.104714 │ 0.995778 │ 0.671143  │ 0.410917 │ 0.547495  │
│ 9   │ 0.373802 │ 0.147139 │ 0.425674  │ 0.893018 │ 0.937846  │
│ 10  │ 0.26242  │ 0.30546  │ 0.750857  │ 0.196515 │ 0.0470533 │
```

The benefit with dropping this constructor is that it would allow more inputs to fall back to the tables constructor, including array types that implement the tables interface like [AxisKeys.jl](https://github.com/mcabbott/AxisKeys.jl). Currently, you need to special case `DataFrame(A)` to `DataFrame(Tables.columns(A))` only when `A` is 2 dimensional.

### PR DESCRIPTION
**Pull request by bkamins**
Fixes #2433

Related https://github.com/JuliaData/Tables.jl/pull/198, https://github.com/mcabbott/AxisKeys.jl/issues/23 and https://github.com/mcabbott/AxisKeys.jl/issues/22.

@rofinn - can you please check if this gives you what you wanted? (note that there is a small issue with column copying but I think it should be resolved on AxisKeys.jl side as noted in https://github.com/mcabbott/AxisKeys.jl/issues/23, as this is a general design principle we use to avoid duplication of copying)

### CODE REVIEW

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -262,8 +264,15 @@ DataFrame(columns::NTuple{N, AbstractVector}; copycols::Bool=true) where {N} =
     DataFrame(collect(AbstractVector, columns), gennames(length(columns)),
               copycols=copycols)
 
-DataFrame(columns::AbstractMatrix,
-          cnames::AbstractVector{Symbol} = gennames(size(columns, 2));
+DataFrame(columns::T) where {T <: AbstractMatrix} =
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:267]: Use the long function form?

File: src/dataframe/dataframe.jl [hunk 2]
```diff
@@ -262,8 +264,15 @@ DataFrame(columns::NTuple{N, AbstractVector}; copycols::Bool=true) where {N} =
...
+    if Tables.istable(T)
+        fromcolumns(Tables.columns(x), collect(Symbol, Tables.columnnames(cols)),
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:269]:
```suggestion
        fromcolumns(Tables.columns(columns),
                    collect(Symbol, Tables.columnnames(columns)),
```

File: src/other/tables.jl [hunk 3]
```diff
@@ -43,7 +43,7 @@ fromcolumns(x, names; copycols::Bool=true) =
               copycols=copycols)
 
 function DataFrame(x::T; copycols::Bool=true) where {T}
-    if !Tables.istable(x)
+    if !Tables.istable(T)
```
Comment (nalimilan) [inline src/other/tables.jl:46]: Is this just to work around the mentioned bug?

File: src/dataframe/dataframe.jl (hunk 2 above)
Comment (rofinn) [reply src/dataframe/dataframe.jl:269]: ☝️ Hit this trying to test out the branch

File: src/dataframe/dataframe.jl (hunk 1 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl:267]: OK

File: src/other/tables.jl (hunk 3 above)
Comment (bkamins) [reply src/other/tables.jl:46]: No - I just think it is cleaner to dispatch on type and we it available as this is the "intended" way to use `istable` (but in particular - while we have the bug - it resolves it)

**Comment by nalimilan** [comment]
Is there a way to test this?

**Comment by bkamins** [comment]
Since CI fails :). I will write a test for this (simply a custom type needs to be defined - I will add it)

**Comment by rofinn** [comment]
FWIW, I can confirm that this works as I'd expect now. Thanks!

**Comment by bkamins** [comment]
OK - added more tests and I had to add `copycols` kwarg there to avoid confusing error messages.

**Comment by bkamins** [comment]
Closing this PR, as we will use another approach. I will open a fresh one.

**Pull request closed without merging**